pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/loader/invertedindex/InvertedIndexHandler.java

```
@@ -42,95 +42,47 @@
-    Set<String> invertedIndexColumns = new HashSet<>();
-    if (indexConfig == null) {
-      return invertedIndexColumns;
-    }
-
-    Set<String> invertedIndexColumnsFromConfig = indexConfig.getInvertedIndexColumns();
-    for (String column : invertedIndexColumnsFromConfig) {
+  private final File _indexDir;
+  private final SegmentDirectory.Writer _segmentWriter;
+  private final String _segmentName;
+  private final SegmentVersion _segmentVersion;
+  private final Set<ColumnMetadata> _invertedIndexColumns = new HashSet<>();
+
+  public InvertedIndexHandler(@Nonnull File indexDir, @Nonnull SegmentMetadataImpl segmentMetadata,
+      @Nonnull IndexLoadingConfig indexLoadingConfig, @Nonnull SegmentDirectory.Writer segmentWriter) {
+    _indexDir = indexDir;
+    _segmentWriter = segmentWriter;
+    _segmentName = segmentMetadata.getName();
+    _segmentVersion = SegmentVersion.valueOf(segmentMetadata.getVersion());
+
+    // Do not create inverted index for sorted column
+    for (String column : indexLoadingConfig.getInvertedIndexColumns()) {
       ColumnMetadata columnMetadata = segmentMetadata.getColumnMetadataFor(column);
       if (columnMetadata != null && !columnMetadata.isSorted()) {
-        invertedIndexColumns.add(column);
+        _invertedIndexColumns.add(columnMetadata);
       }
     }
+  }
 
-    return invertedIndexColumns;
+  public void createInvertedIndices() throws IOException {
+    for (ColumnMetadata columnMetadata : _invertedIndexColumns) {
+      createInvertedIndexForColumn(columnMetadata);
+    }
   }
 
-  private void createInvertedIndexForColumn(ColumnMetadata columnMetadata)
-      throws IOException {
-    // Validation check
+  private void createInvertedIndexForColumn(ColumnMetadata columnMetadata) throws IOException {
     String column = columnMetadata.getColumnName();
-    int totalDocs = columnMetadata.getTotalDocs();
-    if (totalDocs > OffHeapBitmapInvertedIndexCreator.MAX_NUM_ENTRIES) {
```

Are these conditions possible for segments created in hadoop prior to this change?

---

Yes, and we should fail the job instead of skip the inverted index creation

---

But those jobs have already run for past segments, right? Unless I am missing something.